reference/src/representation/structs-and-tuples.md

```diff
@@ -0,0 +1,258 @@
+- `#[repr(transparent)]` -- request that a single-field struct be
+  treated "as if" it were an instance of its field type when passed as
+  an argument
+
+### Default layout ("repr rust")
+
+The default layout of structs is undefined and subject to change
+between compiler revisions. We further do not guarantee that two
+structs with different names (but the same field types) will be laid
+out in the same way (for example, the hypothetical struct representing
+tuples ). Finally, the presence or absence of generics can make a
+difference (e.g., `struct Foo { x: u16, y: u32 }` and `struct Foo<T> {
+x: u16, y: T }` where `T = u32` are not guaranteed to be identical),
+owing to the possibility of unsizing coercions.
+
+**Compiler's current behavior.** As of the time of this writing, the
+compiler will reorder struct fields to minimize the overall size of
+the struct (and in particular to eliminate padding due to alignment
+restrictions). The final field, however, is not reordered if an
+unsizing coercion may be applied.
+
+### C-compatible layout ("repr C")
+
+For structs tagged `#[repr(C)]`, the compiler will apply a C-like
+layout scheme (see section 6.7.2.1 of the [C17 specification][C17] for
+a detailed write-up):
+
+[C17]: http://www.open-std.org/jtc1/sc22/wg14/www/abq/c17_updated_proposed_fdis.pdf
+
+- Field order is preserved.
+- The first field begins at offset 0.
+- Assuming the struct is not packed, each field's offset is aligned to
+  the natural alignment for that field's type, possibly creating
+  unused padding bits.
+- The total size of the struct is rounded up to its overall alignemnt.  
+
+The intention is that if one has a set of C struct declarations and a
+corresponding set of Rust struct declarations, all of which are tagged
+with `#[repr(C)]`, then the layout of those structs will all be
+identical. Note that this setup implies that none of the structs in
+question can contain any `#[repr(Rust)]` structs (or Rust tuples), as
+those would have no corresponding C struct declaration -- as
+`#[repr(Rust)]` types have undefined layout, you cannot safely declare
+their layout in a C program.
+
+See also the notes on ABI compatibility under the section on `#[repr(transparent)]`.
+
+### Fixed alignment
+
+The `#[repr(align(N))]` attribute may be used to raise the alignment
+of a struct, as described in [The Rust Reference][TRR-align].
+
+[TRR-align]: (https://doc.rust-lang.org/stable/reference/type-layout.html#the-align-representation).
+
+### Packed layout
+
+The `#[repr(packed)]` attribute may be used to remove all padding from
+the struct. The resulting fields may not fall at properly aligned
+boundaries in memory. This makes it unsafe to create a Rust reference
+(`&T` or `&mut T`) to those fields, as the compiler requires that all
+reference values must always be aligned (so that it can use more
+efficient load/store instructions at runtime). See the [Rust reference
+for more details][TRR-packed].
```

Comment:
Should mention `#[repr(packed(N))]` (or is it `#[repr(pack(N))]`?) - in general, packing doesn't "remove all padding", but imposes a maximum limit on the alignment guaranteed of each fields.

E.g. in a `#[repr(pack(4))]` struct, fields with alignment `1`, `2` or `4` are placed as usual, at an offset multiple of that alignment, but fields with *larger* alignment than `4` don't get anything better than a multiple of `4`.

Reply:
OK, I wasn't aware of that generalization.